group-coordinator/src/main/java/org/apache/kafka/coordinator/group/generic/GenericGroup.java

```diff
@@ -0,0 +1,958 @@
+     * a static member.
+     *
+     * @param groupInstanceId the group instance id.
+     * @return true if a static member with the group instance id exists,
+     *         false otherwise.
+     */
+    public boolean hasStaticMember(String groupInstanceId) {
+        return staticMembers.containsKey(groupInstanceId);
+    }
+
+    /**
+     * Get member id of a static member that matches the given group
+     * instance id.
+     *
+     * @param groupInstanceId the group instance id.
+     * @return the static member if it exists.
+     */
+    public String staticMemberId(String groupInstanceId) {
+        return staticMembers.get(groupInstanceId);
+    }
+
+    /**
+     * @return members who have yet to rejoin during the
+     *         join group phase.
+     */
+    public Map<String, GenericGroupMember> notYetRejoinedMembers() {
+        Map<String, GenericGroupMember> notYetRejoinedMembers = new HashMap<>();
+        members.values().forEach(member -> {
+            if (!member.isAwaitingJoin()) {
+                notYetRejoinedMembers.put(member.memberId(), member);
+            }
+        });
+        return notYetRejoinedMembers;
+    }
+
+    /**
+     * @return whether all members have joined.
+     */
+    public boolean hasAllMembersJoined() {
+        return members.size() == numMembersAwaitingJoinResponse && pendingJoinMembers.isEmpty();
+    }
+
+    /**
+     * @return the ids of all members in the group.
+     */
+    public Set<String> allMemberIds() {
+        return members.keySet();
+    }
+
+    /**
+     * @return all static members in the group.
+     */
+    public Set<String> allStaticMemberIds() {
+        return staticMembers.keySet();
+    }
+
+    // For testing only.
+    Set<String> allDynamicMemberIds() {
+        Set<String> dynamicMemberSet = new HashSet<>(allMemberIds());
+        staticMembers.values().forEach(dynamicMemberSet::remove);
+        return dynamicMemberSet;
+    }
+
+    /**
+     * @return number of members waiting for a join group response.
+     */
+    public int numAwaiting() {
+        return numMembersAwaitingJoinResponse;
+    }
+
+    /**
+     * @return all members.
+     */
+    public Collection<GenericGroupMember> allMembers() {
+        return members.values();
+    }
+
+    /**
+     * @return the group's rebalance timeout in milliseconds.
+     *         It is the max of all member's rebalance timeout.
+     */
+    public int rebalanceTimeoutMs() {
+        int maxRebalanceTimeoutMs = 0;
+        for (GenericGroupMember member : members.values()) {
+            maxRebalanceTimeoutMs = Math.max(maxRebalanceTimeoutMs, member.rebalanceTimeoutMs());
+        }
+        return maxRebalanceTimeoutMs;
+    }
+
+    /**
+     * Generate a member id from the given client and group instance ids.
+     *
+     * @param clientId the client id.
+     * @param groupInstanceId the group instance id.
+     * @return the generated id.
+     */
```

Comment:
Should we indent this better? 
     * @param clientId        the client id.
     * @param groupInstanceId the group instance id.
     * @return                the generated id.